### PR TITLE
types: fix incompatible implementation of jsonpath extraction

### LINKED ISSUE #30352: Incompatible implementation of jsonpath extraction
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
/* t */ drop table if exists t;
/* t */ create table t (a varchar(100));
/* t */ insert into t values ('{"labels":[{"name":"foo"}]}'),('{"labels":[{"name":"foo"},{"name":"bar"}]}');
/* t */ select a->'$.labels[*].name' from t;
```

### 2. What did you expect to see? (Required)

```sql
/* t */ drop table if exists t;
-- t >> 0 rows affected
/* t */ create table t (a varchar(100));
-- t >> 0 rows affected
/* t */ insert into t values ('{"labels":[{"name":"foo"}]}'),('{"labels":[{"name":"foo"},{"name":"bar"}]}');
-- t >> 2 rows affected
/* t */ select a->'$.labels[*].name' from t;
-- t >> +-----------------------+
-- t    | a->'$.labels[*].name' |
-- t    +-----------------------+
-- t    | ["foo"]               |
-- t    | ["foo", "bar"]        |
-- t    +-----------------------+
```

### 3. What did you see instead (Required)

```sql
/* t */ drop table if exists t;
-- t >> 0 rows affected
/* t */ create table t (a varchar(100));
-- t >> 0 rows affected
/* t */ insert into t values ('{"labels":[{"name":"foo"}]}'),('{"labels":[{"name":"foo"},{"name":"bar"}]}');
-- t >> 2 rows affected
/* t */ select a->'$.labels[*].name' from t;
-- t >> +-----------------------+
-- t    | a->'$.labels[*].name' |
-- t    +-----------------------+
-- t    | "foo"                 | <- not an array
-- t    | ["foo", "bar"]        |
-- t    +-----------------------+
```

### 4. What is your TiDB version? (Required)

master (6475e8949cf4bc4cddfc4cab7dfaf4f6b1bb9fa9)

### PR DESCRIPTION
**Pull request by mengxin9014**
### What problem does this PR solve?
fix incompatible implementation of jsonpath extraction

Issue Number: close #30352

Problem Summary:
```sql
/* t */ drop table if exists t;
-- t >> 0 rows affected
/* t */ create table t (a varchar(100));
-- t >> 0 rows affected
/* t */ insert into t values ('{"labels":[{"name":"foo"}]}'),('{"labels":[{"name":"foo"},{"name":"bar"}]}');
-- t >> 2 rows affected
/* t */ select a->'$.labels[*].name' from t;

except:
-- t >> +-----------------------+
-- t    | a->'$.labels[*].name' |
-- t    +-----------------------+
-- t    | ["foo"]               |
-- t    | ["foo", "bar"]        |
-- t    +-----------------------+

actual:
-- t >> +-----------------------+
-- t    | a->'$.labels[*].name' |
-- t    +-----------------------+
-- t    | "foo"                 | <- not an array
-- t    | ["foo", "bar"]        |
-- t    +-----------------------+
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by zanmato1984 — COMMENTED** [review]
Please add some test cases.

File: types/json/binary_test.go [hunk 1]
```diff
@@ -347,6 +347,7 @@ func TestBinaryJSONMerge(t *testing.T) {
 
 func mustParseBinaryFromString(t *testing.T, s string) BinaryJSON {
 	result, err := ParseBinaryFromString(s)
+	print(s)
```
Comment (xhebox) [inline types/json/binary_test.go:350]: Useless debugging?

**Review by xhebox — COMMENTED** [review]
LGTM

File: types/json/binary_test.go (hunk 1 above)
Comment (mengxin9014) [reply types/json/binary_test.go]: > Useless debugging?

I forgot to delete it. 
deleted.

**Review by xhebox — APPROVED** [review]

**Comment by mengxin9014** [comment]
> LGTM, but maybe add `TestIssue30352` as another reviewer requested.

The tests are included in [binary_test.go](https://github.com/pingcap/tidb/pull/35320/files#diff-94e7d7dbe4dff695f8f6cb8f60025b9d34144d7e28016cf58f73e1f9629a7584)

**Comment by xhebox** [comment]
> > LGTM, but maybe add `TestIssue30352` as another reviewer requested.
> 
> The tests are included in [binary_test.go](https://github.com/pingcap/tidb/pull/35320/files#diff-94e7d7dbe4dff695f8f6cb8f60025b9d34144d7e28016cf58f73e1f9629a7584)

I knew, so I approved. But he requested. Anyway, you can re-request reviews from him or others.

**Review by zanmato1984 — APPROVED** [review]
LGTM

**Comment by mengxin9014** [comment]
> 

ok

**Pull request merged**